libff/algebra/scalar_multiplication/test_multiexp.cpp

```diff
@@ -135,41 +147,243 @@ template<typename GroupT, multi_exp_method Method> void test_multiexp_inner()
 }
 
-TEST(MultiExpTest, TestMultiExpAccumulateBuckets)
+template<typename GroupT> void test_multiexp_accumulate_buckets()
 {
-    test_multiexp_accumulate_buckets<alt_bn128_G1>(4);
-    test_multiexp_accumulate_buckets<alt_bn128_G1>(16);
-    test_multiexp_accumulate_buckets<alt_bn128_G1>(128);
+    test_multiexp_accumulate_buckets_group<GroupT, multi_exp_base_form_normal>(
+        4);
+    test_multiexp_accumulate_buckets_group<GroupT, multi_exp_base_form_normal>(
+        16);
+    test_multiexp_accumulate_buckets_group<GroupT, multi_exp_base_form_normal>(
+        128);
+    test_multiexp_accumulate_buckets_group<GroupT, multi_exp_base_form_special>(
+        4);
+    test_multiexp_accumulate_buckets_group<GroupT, multi_exp_base_form_special>(
+        16);
+    test_multiexp_accumulate_buckets_group<GroupT, multi_exp_base_form_special>(
+        128);
 }
 
-TEST(MultiExpTest, TestMultiExpSignedDigitsRound)
+template<typename GroupT> void test_multiexp_signed_digits_round()
 {
-    test_multiexp_signed_digits_round<alt_bn128_G1, false>(0);
-    test_multiexp_signed_digits_round<alt_bn128_G1, true>(0);
-    test_multiexp_signed_digits_round<alt_bn128_G1, false>(1);
-    test_multiexp_signed_digits_round<alt_bn128_G1, true>(1);
-    test_multiexp_signed_digits_round<alt_bn128_G1, false>(2);
-    test_multiexp_signed_digits_round<alt_bn128_G1, true>(2);
-    test_multiexp_signed_digits_round<alt_bn128_G1, false>(3);
-    test_multiexp_signed_digits_round<alt_bn128_G1, true>(3);
+    test_multiexp_signed_digits_round_group<GroupT, multi_exp_base_form_normal>(
+        0);
+    test_multiexp_signed_digits_round_group<
+        GroupT,
+        multi_exp_base_form_special>(0);
+    test_multiexp_signed_digits_round_group<GroupT, multi_exp_base_form_normal>(
+        1);
+    test_multiexp_signed_digits_round_group<
+        GroupT,
+        multi_exp_base_form_special>(1);
+    test_multiexp_signed_digits_round_group<GroupT, multi_exp_base_form_normal>(
+        2);
+    test_multiexp_signed_digits_round_group<
+        GroupT,
+        multi_exp_base_form_special>(2);
+    test_multiexp_signed_digits_round_group<GroupT, multi_exp_base_form_normal>(
+        3);
+    test_multiexp_signed_digits_round_group<
+        GroupT,
+        multi_exp_base_form_special>(3);
 }
 
-TEST(MultiExpTest, TestMultiExpInner)
+// template<typename GroupT, multi_exp_method Method, multi_exp_base_form
+// baseForm>
+template<typename GroupT, multi_exp_method Method, multi_exp_base_form BaseForm>
+void test_multi_exp_config(size_t num_elements)
 {
-    test_multiexp_inner<alt_bn128_G1, multi_exp_method_BDLO12_signed_mixed>();
+    using Field = typename GroupT::scalar_field;
+
+    std::vector<GroupT> base_elements;
+    base_elements.reserve(num_elements);
+    std::vector<Field> scalars;
+    scalars.reserve(num_elements);
+
+    // Create values for sum [n]*[1]_G + [n-1]*[2]_G + ... + [1]*[n]_G
+    // where [x]_G is encoding of x in G. The result should then be
```

Comment:
If `[x]_G` is the notation for group encoding, then what is the notation `[x]` (without the subscript)? Shouldn't we get rid of these brackets here?

Reply:
`[a]P` is commonly used to explicitly denote multiplication of a group element by a scalar.  Removed to avoid any confusion with `[.]_G`.

Reply:
> [a]P is commonly used to explicitly denote multiplication of a group element by a scalar. Removed to avoid any confusion with [.]_G.

There are plently of notations for this type of things. In general group elements are denoted by Capital letters (generally P, Q, R etc.) while scalars are denoted in small caps starting from (a, b etc). This alone is sufficient most of the time as notation to disambiguate things like `aP`. Sometimes people use the bracket notation `[a]` to represent `a*P` (or the encoding of scalar a in the group with generator P). In general I don't have strong preferences over notations - other than trying to stick as mush as possible to the ones we commonly use, just for simplicity - as long as things are properly documented.